snuba/datasets/plans/translator/visitor.py

```diff
@@ -0,0 +1,158 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from typing import Generic, Optional, Sequence, TypeVar
+
+from snuba.query.expressions import (
+    Argument,
+    Column,
+    CurriedFunctionCall,
+    Expression,
+    ExpressionVisitor,
+    FunctionCall,
+    Lambda,
+    Literal,
+    SubscriptableReference,
+)
+
+TExpOut = TypeVar("TExpOut")
+
+
+@dataclass(frozen=True)
+class TranslationRules(Generic[TExpOut]):
+    """
+    Represents the set of rules to be used to configure a MappingExpressionTranslator.
+    It encapsulates a sequence or rules for each AST node type.
+    It provides a concat method to combine multiple sets of rules.
+
+    see MappingExpressionTranslator for more context
+    """
+
+    literals: Sequence[ExpressionMapper[Literal, TExpOut]] = field(default_factory=list)
+    columns: Sequence[ExpressionMapper[Column, TExpOut]] = field(default_factory=list)
+    subscriptables: Sequence[ExpressionMapper[SubscriptableReference, TExpOut]] = field(
+        default_factory=list
+    )
+    functions: Sequence[ExpressionMapper[FunctionCall, TExpOut]] = field(
+        default_factory=list
+    )
+    curried_functions: Sequence[ExpressionMapper[CurriedFunctionCall, TExpOut]] = field(
+        default_factory=list
+    )
+    arguments: Sequence[ExpressionMapper[Argument, TExpOut]] = field(
+        default_factory=list
+    )
+    lambdas: Sequence[ExpressionMapper[Lambda, TExpOut]] = field(default_factory=list)
+
+    def concat(self, spec: TranslationRules[TExpOut]) -> TranslationRules[TExpOut]:
+        return TranslationRules(
+            literals=[*self.literals, *spec.literals],
+            columns=[*self.columns, *spec.columns],
+            subscriptables=[*self.subscriptables, *spec.subscriptables],
+            functions=[*self.functions, *spec.functions],
+            curried_functions=[*self.curried_functions, *spec.curried_functions],
+            arguments=[*self.arguments, *spec.arguments],
+            lambdas=[*self.lambdas, *spec.lambdas],
+        )
+
+
+TExpIn = TypeVar("TExpIn", bound=Expression)
+
+
+class ExpressionMapper(ABC, Generic[TExpIn, TExpOut]):
+    """
+    One translation rule used by the MappingExpressionTranslator to translate an
+    expression.
+
+    see MappingExpressionTranslator for more context
+    """
+
+    @abstractmethod
+    def attemptMap(
+        self, expression: TExpIn, children_translator: ExpressionVisitor[TExpOut]
+    ) -> Optional[TExpOut]:
+        """
+        Translates an expression if this rule matches such expression. If not, it
+        returns None.
+
+        It receives an instance of a translating visitor to take care of children
+        that this rule is not capable, or not supposed to, translate. These node
+        should be delegated to the children_translator.
```

Comment:
Some additional context on why we need to pass the translator to an ExpressionMapper and delegate to the mapper the responsibility to apply the translator to its children:

in the original AST transform method, the FunctionCall node guides this process: it first applies the transformation lambda to all children (parameters), then it replaces the old parameter with the new one on the FunctionaCall node and then it applies the lambda to the new node and returns the result.

I originally wanted to do something like this where the visitor ensures first we translate the parameter of a function then we find a matcher for the function itself with the translated parameters.
Unfortunately the translation produces a different data type than the one it took in input (they are not Expressions since they are the equivalent of the clickhouse query) so that simple process breaks. We cannot pack the logical pre-translation FunctionCall with the post-translation parameters and find a mapper for that. 
Thus we need the mapper to return a fully translated subtree (that type checks properly), and we provide a translator to the mapper so it does not have to figure out how to translate its children.